File: app/src/main/java/com/aivoice/translate/Activity/IntroActivity.java
package com.aivoice.translate.Activity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.Html;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.viewpager.widget.PagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.aivoice.translate.R;
import com.aivoice.translate.base.BaseActivity;

import butterknife.BindView;
import butterknife.OnClick;

public class IntroActivity extends BaseActivity {

    @BindView(R.id.mViewPager)
    ViewPager mViewPager;
    @BindView(R.id.mLLTop)
    LinearLayout mLLTop;
    @BindView(R.id.mTxtSkip)
    TextView mTxtSkip;
    @BindView(R.id.mLayoutDots)
    LinearLayout mLayoutDots;
    @BindView(R.id.mTxtNext)
    TextView mTxtNext;
    private TextView[] dots;
    private int[] layouts;
    private MyViewPagerAdapter myViewPagerAdapter;

    @Override
    protected int getLayoutId() {
        return R.layout.activity_intro;
    }

    @Override
    protected void initView(Bundle savedInstanceState) {
        if (!prefManager.isFirstTimeLaunch()) {
            launchHomeScreen();
            finish();
        }

        layouts = new int[]{
                R.layout.layout_into1,
                R.layout.layout_into2,
                R.layout.layout_into3};

        addBottomDots(0);

        myViewPagerAdapter = new MyViewPagerAdapter();
        mViewPager.setAdapter(myViewPagerAdapter);
        mViewPager.addOnPageChangeListener(viewPagerPageChangeListener);
    }


    @OnClick({R.id.mTxtNext, R.id.mTxtSkip})
    public void onViewClicked(View view) {
        switch (view.getId()) {
            case R.id.mTxtNext:
                int current = getItem(+1);
                if (current < layouts.length) {
                    // move to next screen
                    mViewPager.setCurrentItem(current);
                } else {
                    launchHomeScreen();
                }
                break;
            case R.id.mTxtSkip:
                launchHomeScreen();
                break;
        }
    }

    private void addBottomDots(int currentPage) {

        dots = new TextView[layouts.length];

        int[] colorsActive = getResources().getIntArray(R.array.array_dot_active);
        int[] colorsInactive = getResources().getIntArray(R.array.array_dot_inactive);

        mLayoutDots.removeAllViews();
        for (int i = 0; i < dots.length; i++) {
            dots[i] = new TextView(this);
            dots[i].setText(Html.fromHtml("&#8226;"));
            dots[i].setTextSize(28);
            dots[i].setTextColor(colorsInactive[currentPage]);
            mLayoutDots.addView(dots[i]);
        }

        if (dots.length > 0)
            dots[currentPage].setTextColor(colorsActive[currentPage]);
    }

    private int getItem(int i) {
        return mViewPager.getCurrentItem() + i;
    }

    private void launchHomeScreen() {
        prefManager.setFirstTimeLaunch(false);
        startActivity(new Intent(IntroActivity.this, HomeActivity.class));
        finish();
    }

    ViewPager.OnPageChangeListener viewPagerPageChangeListener = new ViewPager.OnPageChangeListener() {

        @Override
        public void onPageSelected(int position) {
            if (position == 2) {
                mTxtSkip.setVisibility(View.GONE);
            } else {
                mTxtSkip.setVisibility(View.VISIBLE);
            }
            addBottomDots(position);
        }

        @Override
        public void onPageScrolled(int arg0, float arg1, int arg2) {

        }

        @Override
        public void onPageScrollStateChanged(int arg0) {

        }
    };

    public class MyViewPagerAdapter extends PagerAdapter {
        private LayoutInflater layoutInflater;

        public MyViewPagerAdapter() {
        }

        @Override
        public Object instantiateItem(ViewGroup container, int position) {
            layoutInflater = (LayoutInflater) getSystemService(Context.LAYOUT_INFLATER_SERVICE);

            View view = layoutInflater.inflate(layouts[position], container, false);
            container.addView(view);

            return view;
        }

        @Override
        public int getCount() {
            return layouts.length;
        }

        @Override
        public boolean isViewFromObject(View view, Object obj) {
            return view == obj;
        }


        @Override
        public void destroyItem(ViewGroup container, int position, Object object) {
            View view = (View) object;
            container.removeView(view);
        }
    }
}

File: app/src/main/java/com/texttranslate/voiceimage/IntroActivity.kt
package com.texttranslate.voiceimage

import android.content.Intent
import android.os.Bundle
import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.viewpager.widget.PagerAdapter
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.texttranslate.voiceimage.base.PrefManager
import com.texttranslate.voiceimage.databinding.ActivityIntroBinding

class IntroActivity : AppCompatActivity(), View.OnClickListener {
    private var layouts: IntArray?= null
    var binding: ActivityIntroBinding? = null
    var prefManager: PrefManager? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityIntroBinding.inflate(
            layoutInflater
        )
        setContentView(binding!!.getRoot())
        prefManager = PrefManager(this)
        if (!prefManager!!.isFirstTimeLaunch) {
            launchHomeScreen()
            finish()
        }
        layouts = intArrayOf(
            R.layout.layout_into1,
            R.layout.layout_into2,
            R.layout.layout_into3
        )
        addBottomDots(0)
        val myViewPagerAdapter = MyViewPagerAdapter()
        binding!!.mViewPager.setAdapter(myViewPagerAdapter)
        binding!!.mViewPager.addOnPageChangeListener(viewPagerPageChangeListener)
        binding!!.mTxtNext.setOnClickListener(this)
        binding!!.mTxtSkip.setOnClickListener(this)
    }

    override fun onClick(view: View) {
        if (view === binding!!.mTxtNext) {
            val current = item
            if (current < layouts!!.size) {
                // move to next screen
                binding!!.mViewPager.setCurrentItem(current)
            } else {
                launchHomeScreen()
            }
        } else if (view === binding!!.mTxtSkip) {
            launchHomeScreen()
        }
    }

    private fun addBottomDots(currentPage: Int) {
        val dots = arrayOfNulls<TextView>(layouts!!.size)
        val colorsActive = getResources().getIntArray(R.array.array_dot_active)
        val colorsInactive = getResources().getIntArray(R.array.array_dot_inactive)
        binding!!.mLayoutDots.removeAllViews()
        for (i in dots.indices) {
            dots[i] = TextView(this)
            dots[i]!!.text = Html.fromHtml("&#8226;")
            dots[i]!!.textSize = 28f
            dots[i]!!.setTextColor(colorsInactive[currentPage])
            binding!!.mLayoutDots.addView(dots[i])
        }
        if (dots.size > 0) dots[currentPage]!!.setTextColor(colorsActive[currentPage])
    }

    private val item: Int
        private get() = binding!!.mViewPager.currentItem + 1

    private fun launchHomeScreen() {
        prefManager!!.isFirstTimeLaunch = false
        startActivity(Intent(this@IntroActivity, HomeActivity::class.java))
        finish()
    }

    var viewPagerPageChangeListener: OnPageChangeListener = object : OnPageChangeListener {
        override fun onPageSelected(position: Int) {
            if (position == 2) {
                binding!!.mTxtSkip.visibility = View.GONE
            } else {
                binding!!.mTxtSkip.visibility = View.VISIBLE
            }
            addBottomDots(position)
        }

        override fun onPageScrolled(arg0: Int, arg1: Float, arg2: Int) {}
        override fun onPageScrollStateChanged(arg0: Int) {}
    }

    inner class MyViewPagerAdapter : PagerAdapter() {
        override fun instantiateItem(container: ViewGroup, position: Int): Any {
            val layoutInflater = getSystemService(LAYOUT_INFLATER_SERVICE) as LayoutInflater
            val view = layoutInflater.inflate(layouts!![position], container, false)
            container.addView(view)
            return view
        }

        override fun getCount(): Int {
            return layouts!!.size
        }

        override fun isViewFromObject(view: View, obj: Any): Boolean {
            return view === obj
        }

        override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
            val view = `object` as View
            container.removeView(view)
        }
    }
}